presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,261 @@
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class ShardOrganizationManager
+{
+    private static final Logger log = Logger.get(ShardOrganizationManager.class);
+
+    private final ScheduledExecutorService discoveryService = newScheduledThreadPool(1, daemonThreadsNamed("shard-organization-discovery"));
+    private final AtomicBoolean started = new AtomicBoolean();
+    private final AtomicBoolean shutdown = new AtomicBoolean();
+
+    private final IDBI dbi;
+    private final MetadataDao metadataDao;
+    private final ShardOrganizerDao shardOrganizerDao;
+    private final ShardManager shardManager;
+
+    private final boolean enabled;
+    private final Duration discoveryInterval;
+
+    private final String currentNodeIdentifier;
+    private final ShardOrganizer organizer;
+
+    @Inject
+    public ShardOrganizationManager(@ForMetadata IDBI dbi, NodeManager nodeManager, ShardManager shardManager, ShardOrganizer organizer, StorageManagerConfig config)
+    {
+        this(dbi,
+                nodeManager.getCurrentNode().getNodeIdentifier(),
+                shardManager,
+                organizer,
+                config.isOrganizationEnabled(),
+                config.getOrganizationInterval());
+    }
+
+    public ShardOrganizationManager(
+            IDBI dbi,
+            String currentNodeIdentifier,
+            ShardManager shardManager,
+            ShardOrganizer organizer,
+            boolean enabled,
+            Duration discoveryInterval)
+    {
+        this.dbi = requireNonNull(dbi, "dbi is null");
+        this.metadataDao = onDemandDao(dbi, MetadataDao.class);
+        this.shardOrganizerDao = onDemandDao(dbi, ShardOrganizerDao.class);
+
+        this.organizer = requireNonNull(organizer, "organizer is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.currentNodeIdentifier = requireNonNull(currentNodeIdentifier, "currentNodeIdentifier is null");
+
+        this.enabled = enabled;
+        this.discoveryInterval = requireNonNull(discoveryInterval, "discoveryInterval is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!enabled) {
+            return;
+        }
+
+        if (!started.getAndSet(true)) {
+            startDiscovery();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        discoveryService.shutdown();
+    }
+
+    private void startDiscovery()
+    {
+        long startDelay = getStartDelay();
+
+        discoveryService.scheduleWithFixedDelay(() -> {
+            if (shutdown.get()) {
+                return;
+            }
+
+            try {
+                // jitter to avoid overloading database
+                long interval = (long) discoveryInterval.convertTo(SECONDS).getValue();
```

Comment:
Using the full interval as the jitter doesn't make sense with the default interval of 7 days, since that would mean things actually run every 10.5 days on average. How about a fixed jitter of `min(1 hour, interval)`?